keps/sig-storage/20191125-bucket-provisioning.md

```diff
@@ -0,0 +1,428 @@
+
+
+```yaml
+apiVersion: cosi.io/v1alpha1
+kind: BucketRequest
+metadata:
+  name:
+  namespace:
+  labels:
+    cosi.io/provisioner: [1]
+  finalizers:
+  - cosi.io/finalizer [2]
+spec:
+  protocol: [3]
+  bucketPrefix: [4]
+  bucketClassName: [5]
+  secretName: [6]
+  bucketName: [7]
+status:
+  phase: [9]
+  conditions: 
+```
+
+1. `labels`: should be added by controller.  Key’s value should be the provisioner name. Characters that do not adhere to [Kubernetes label conventions](https://kubernetes.io/docs/concepts/overview/working-with-objects/labels/#syntax-and-character-set) will be converted to ‘-’.
+1. `finalizers`: should be added by controller to defer BucketRequest deletion until backend deletion ops succeed.
+1. `protocol`: specifies the desired protocol.  One of {“s3”, “gcs”, or “azureBlob”}.
+1. `bucketPrefix`: (Optional) prefix prepended to a randomly generated bucket name, eg. “yosemite-photos-". If empty no prefix is prepended.
+1. `bucketClassName`: Name of the target BucketClass.
+1. `secretName`: (optional) Secret in the BucketRequest's namespace storing credentials to be used by a workload for bucket access.
+1. `bucketName`: Name of the Bucket instance representing the desired storage instance endpoint.  Set by the COSI Controller for new Buckets.  Specified by the user when connecting to existing buckets.
+1. `phase`: 
+   - *Pending*: The controller has detected the new `Bucket` and begun provisioning operations
+   - *Bound*: Provisioning operations have completed and the `Bucket` has been bound to a `Bucket`.
+
+#### Bucket
+
+A cluster-scoped resource representing a storage instance.  A Bucket should contain enough information to enable the requesting client to access the storage instance.  At a minimum, a Bucket must store enough identifying information such in order for drivers to accurately target the storage instance (e.g. during a deletion process).
+
+```yaml
+apiVersion: cosi.io/v1alpha1
+kind: Bucket
+Metadata:
+  name: [1]
+  labels:
+    cosi.io/provisioner: [2]
+  finalizers:
+  - cosi.io/finalizer [3]
+spec:
+  provisioner: [4]
+  releasePolicy: [5]
+  anonymousAccessMode: [6]
+    - private
+    - publicRead
+    - publicReadWrite
+  bucketClassName: [7]
+  permittedNamespaces: [8]
+    - name:
+      uid:
+  protocol: [9]
+    protocolSignature: ""
+    azureBlob: [10]
+      containerName:
+      storageAccount:
+    s3: [11]
+      endpoint:
+      bucketName:
+      region:
+      signatureVersion:
+    gcs: [12]
+      bucketName:
+      privateKeyName:
+      projectId:
+      serviceAccount:
+  parameters: [13]
+status:
+  message: [14]
+  phase: [15]
+  conditions:
+```
+
+1. `name`: For Greenfield, generated in the pattern of `bucket-<my-bucket>-<my-bucket-namespace>`. 
```

Comment:
name format should be more explicit: _bucket-<bucketRequest.ame>-<bucketRequest.namespace>_